src/Marten/Internal/Sessions/EventTracingConnectionLifetime.cs

```diff
@@ -0,0 +1,204 @@
+
+            throw;
+        }
+    }
+
+    public DbDataReader ExecuteReader(NpgsqlBatch batch)
+    {
+        _databaseActivity?.AddEvent(new ActivityEvent(MartenBatchExecutionStarted));
+
+        try
+        {
+            return _innerConnectionLifetime.ExecuteReader(batch);
+        }
+        catch (Exception e)
+        {
+            RecordException(e, MartenBatchExecutionFailed);
+
+            throw;
+        }
+    }
+
+    public Task<DbDataReader> ExecuteReaderAsync(NpgsqlBatch batch, CancellationToken token = default)
+    {
+        _databaseActivity?.AddEvent(new ActivityEvent(MartenBatchExecutionStarted));
+
+        try
+        {
+            return _innerConnectionLifetime.ExecuteReaderAsync(batch, token);
+        }
+        catch (Exception e)
+        {
+            RecordException(e, MartenBatchExecutionFailed);
+
+            throw;
+        }
+    }
+
+    public void ExecuteBatchPages(IReadOnlyList<OperationPage> pages, List<Exception> exceptions)
+    {
+        _databaseActivity?.AddEvent(new ActivityEvent(MartenBatchPagesExecutionStarted));
+
+        try
+        {
+            _innerConnectionLifetime.ExecuteBatchPages(pages, exceptions);
+        }
+        catch (AggregateException e)
+        {
+            RecordExceptions(e, MartenBatchPagesExecutionFailed);
+
+            throw;
+        }
+        catch (Exception e)
+        {
+            RecordException(e, MartenBatchPagesExecutionFailed);
+
+            throw;
+        }
+    }
+
+    public Task ExecuteBatchPagesAsync(IReadOnlyList<OperationPage> pages, List<Exception> exceptions, CancellationToken token)
+    {
+        _databaseActivity?.AddEvent(new ActivityEvent(MartenBatchPagesExecutionStarted));
+
+        try
+        {
+            return _innerConnectionLifetime.ExecuteBatchPagesAsync(pages, exceptions, token);
+        }
+        catch (AggregateException e)
+        {
+            RecordExceptions(e, MartenBatchPagesExecutionFailed);
+
+            throw;
+        }
+        catch (Exception e)
+        {
+            RecordException(e, MartenBatchPagesExecutionFailed);
+
+            throw;
+        }
+    }
+
+    private void RecordException(Exception exceptionToRecord, string eventName)
+    {
+        var tagsToAdd = new ActivityTagsCollection(new[] { new KeyValuePair<string, object?>(MartenCommandFailedExceptionType, exceptionToRecord.GetType()) });
+        _databaseActivity?.AddEvent(new ActivityEvent(eventName, DateTimeOffset.UtcNow, tagsToAdd));
+    }
+
+    private void RecordExceptions(AggregateException exceptionsToRecord, string eventName)
+    {
+        var innerExceptionTypes = exceptionsToRecord.InnerExceptions.Select(t => t.GetType());
+        var tagsToAdd = new ActivityTagsCollection(new[] { new KeyValuePair<string, object?>(MartenCommandFailedExceptionTypes, innerExceptionTypes) });
+        _databaseActivity?.AddEvent(new ActivityEvent(eventName, DateTimeOffset.UtcNow, tagsToAdd));
+    }
```

Comment:
I think it'd be more useful to align with OTEL standards in regards to exception reporting. Consider vendoring `RecordException` from OTEL.NET: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Trace/ActivityExtensions.cs#L81 into `MartenTracing`

AggregateExceptions can have the inner exceptions be individually recorded then.

Reply:
That makes complete sense. The only question I have is whether we want to take a dependency on the OTel API NuGet package. From what I can see that doesn't depend on any other OTel packages, just the DiagnosticSource package. It would probably make sense if we are going to invest in OTel as a way of observing Marten. @jeremydmiller, what do you think?

Reply:
@jeremydmiller  That was my main concern. There are stable versions of the API package, [version 1.8 was just released](https://www.nuget.org/packages/OpenTelemetry.Api).

I'll bring the RecordException method from this package in to Marten later today/tomorrow .

Reply:
@Hawxy  I'll just bring that method and any required constants in then.
Should we be calling Stop on the activity we start in the dispose method?

Reply:
>  Should we be calling Stop on the activity we start in the dispose method?

When the activity is disposed, it stops automatically / internally. So just make sure that the activity gets disposed (used within `using`, etc.).

Reply:
I'm fine with the dependency then. Thanks for all the work on this!

Reply:
@jeremydmiller  OK, I’ll pull in the OpenTelemetry.API package at the latest stable version, assuming that’s what you meant?
Do we want any automated tests in a separate test project, I think I can automate the Prometheus stuff using Docker and TestContainers.

Reply:
Unit tests can be done with the `ActivityListener` from .NET.

OTel setup for exporting is IMO out of scope for Marten, and documented on their side. So I'd just go w/ some basic units tests for activities / tracing, and don't see a need to mess around with Prometheus (actually I think Jaeger would be better for tracing, as Prometheus isn't made for this).

Reply:
https://github.com/JasperFx/alba/blob/master/src/Alba.Testing/ActivityTests.cs Alba has a basic test for OTEL as an example.